### PR TITLE
Refactor layer to handle minibatch at once

### PR DESCRIPTION
**Pull request by reunanen**
For [Batch Normalization purposes](https://github.com/nyanp/tiny-cnn/issues/147#issuecomment-225482682), refactor layer to handle a minibatch at once.

Not saying that I completely knew what I was doing, so please be critical!

### CODE REVIEW

File: tiny_cnn/network.h [hunk 1]
```diff
@@ -672,10 +672,18 @@ class network
...
+            if (start_index < end_index) { // any data for this thread to process?
+                // process data points in this batch assigned to thread i
+                std::vector<tensor_t> in_batch(&in[start_index], &in[end_index]);
+                std::vector<tensor_t> t_batch(&t[start_index], &t[end_index]);
+                std::vector<tensor_t> t_cost_batch
+                    = t_cost
+                    ? std::vector<tensor_t>(&t_cost[start_index], &t_cost[end_index])
+                    : std::vector<tensor_t>();
+
+                bprop<E>(fprop(in_batch, i), t_batch, i, t_cost_batch);
```
Comment (reunanen) [inline tiny_cnn/network.h]: This `bprop`s only those samples from the minibatch that have been assigned to this thread. I guess this will need to be changed when implementing Batch Normalization, so that the whole minibatch goes to the layer as a single object. The parallelism can then be handled inside each layer. This may be a good thing, because the optimal parallelism strategy may be different for each layer type (at least in theory).

File: tiny_cnn/network.h [hunk 2]
```diff
@@ -732,17 +745,15 @@ class network
...
-            });
-        }
+    void bprop(const std::vector<vec_t>& out, const std::vector<vec_t>& t, int idx, const std::vector<vec_t>& t_cost) {
+        bprop<E>(std::vector<tensor_t>{out}, std::vector<tensor_t>{t}, idx, std::vector<tensor_t>{t_cost});
+    }
 
+    template <typename E>
+    void bprop(const std::vector<tensor_t>& out, const std::vector<tensor_t>& t, int idx, const std::vector<tensor_t>& t_cost) {
+        tensor_t delta = gradient<E>(out, t, t_cost);
         net_.backward(delta, idx);
```
Comment (reunanen) [inline tiny_cnn/network.h]: This too may require some further changes when implementing Batch Normalization.

**Comment by edgarriba** [comment]
@reunanen it looks nice! be aware that soon there will be some changes in the layers API.
Check this PR for device abstraction: https://github.com/nyanp/tiny-cnn/pull/166

File: test/test_network.h [hunk 3]
```diff
@@ -156,26 +156,23 @@ TEST(network, multi_out) {
...
         bool in[2] = { bernoulli(0.5), bernoulli(0.5) };
         label_t expected = (in[0] ^ in[1]) ? 1 : 0;
         data.push_back({ in[0] * 1.0, in[1] * 1.0 });
 
         out.emplace_back(std::vector<vec_t>{
-            {(in[0]&&in[1])*1.0, (in[0]||in[1])*1.0}, // 1st output: train and/or function
-            {(in[0]^in[1])*1.0, (in[0]==in[1])*1.0},  // 2nd output: train xor/eq function
-            {-1, 0.5}                                 // 3rd output: train constants (purpose: make output channel count different from output dimension)
+            {(in[0]&&in[1])*1.0, (in[0]||in[1])*1.0}, // 1st output train and/or function
+            {(in[0]^in[1])*1.0, (in[0]==in[1])*1.0} // 2nd output train xor/eq function
```
Comment (edgarriba) [inline test/test_network.h]: @reunanen please cast the results to `float_t` as done [here](https://github.com/edgarriba/tiny-cnn/blob/8756f3f8a9a98420ae5b69c96d7394866f6c3d6e/test/test_network.h#L166) in order to have a consistent code

Comment (reunanen) [inline test/test_network.h]: Ok. But actually I reverted this change already. Didn't really mean to push it in the first place – just added it to more easily track how the different dimensions are supposed to work.

**Comment by edgarriba** [comment]
@reunanen @nyanp this PR is to master. We should define a quick protocol for development branches

**Comment by reunanen** [comment]
Damn, I only tested with OpenMP enabled. Apparently the tests fail without it. Will try to fix.

**Comment by bhack** [comment]
@nyanp Yes we need to find a policy for branches and pull request in this period of heavy refactoring

**Comment by edgarriba** [comment]
my PR can wait after we can merge this minibatch approach. Otherwise, it will be a pain to adapt it.

**Comment by nyanp** [comment]
@reunanen 
Great step! Did you execute a performance profiling? I'm interested in how this PR will affect the speed of forward/backward pass.

@edgarriba 
Oops I've accidentally merged your first PR. As far as I see it is not so hard to merge this PR into current feat/generic-computational-graph, so I'll do it.

**Comment by edgarriba** [comment]
no worries

@nyanp @reunanen @bhack 
Do you think that with the new API we could implement this minibatches routines inside `layer::forward/backward_pass()` ?

**Comment by nyanp** [comment]
@edgarriba 
Yes, and we'll have more opportunities for improve the training time with minibatch parallelism inside each layers.

**Comment by reunanen** [comment]
Unfortunately still struggling with correctness. @nyanp, good question (re: perf), compared it for the first time now. Seems to be poorer at the moment (about 10% or so); will try to look into that as well.

File: tiny_cnn/layers/dropout_layer.h [hunk 4]
```diff
@@ -134,22 +142,26 @@ class dropout_layer : public layer {
...
     }
 
     virtual void set_worker_count(cnn_size_t worker_count) override {
         Base::set_worker_count(worker_count);
         dropout_layer_worker_storage_.resize(worker_count);
 
         for (dropout_layer_worker_specific_storage& dws : dropout_layer_worker_storage_) {
-            dws.mask_.resize(in_size_);
+            if (dws.mask_.empty()) {
+                dws.mask_.resize(1, std::vector<uint8_t>(in_size_));
```
Comment (nyanp) [inline tiny_cnn/layers/dropout_layer.h]: Size of ```mask_``` allocated here is [1, in_size_], but what we need is [sample_size, in_size_]. Possible solution is one of them: (I think the former is better)

- Add new virtual function ```set_minibatch_count``` into layer classes, and call it from ```network``` at the beginning of training
- Check the size of minibatch in fprop and resize ```mask_``` on demand - It requires a exclusive locking while resizing

**Comment by nyanp** [comment]
@reunanen 
> Seems to be poorer at the moment (about 10% or so); will try to look into that as well.

Ok. I think loss around 10% is acceptable for master branch :) but it's great if you improve it better!

> So even if the tests pass, I don't think everything is correct. Probably better not merge just yet.

Current tests do almost nothing about minibatch, so it would be great if you add/extend a minibatch version of ```gradient_check``` method.

**Comment by edgarriba** [comment]
besides, if we parallelize the for loops for mini-batches I think that it will be a huge boosting

**Comment by bhack** [comment]
For convolution there are differents batched gemm approaches. Make a google search on this topic. /cc @naibaf7

File: tiny_cnn/layers/dropout_layer.h (hunk 4 above)
Comment (reunanen) [inline tiny_cnn/layers/dropout_layer.h]: @nyanp: Correct. For other layer types I took the latter approach, though (see e.g. [here](https://github.com/reunanen/tiny-cnn/blob/80272e443b10c3f8ed53d6d28da054d38aee00b8/tiny_cnn/layers/convolutional_layer.h#L494-L495) – although I think this may be one of the places that are not correct yet, but let's not focus on that right now).

Not sure why it requires a lock, though, as each vector (to be resized) is still worker-specific.

Comment (nyanp) [inline tiny_cnn/layers/dropout_layer.h]: @reunanen
Ah I misunderstood, you are correct. it doesn't require a lock while resizing.

Comment (reunanen) [inline tiny_cnn/layers/dropout_layer.h]: Fixed by resizing in fprop.

**Comment by reunanen** [comment]
Regarding the discussion on perf – after all the point is to be able to do Batch Normalization, which should let us learn more in a single epoch. But maybe this shouldn't be merged to master, before BN actually exists?

**Comment by nyanp** [comment]
@reunanen 
As far as I know, all other major dnn libraries(tensorflow, caffe, mxnet, keras) handle minibatch as this PR - it is a good sign for us. Since we may be heading in the right direction regardless of BN, I can merge this PR without BN implementation.

Could you add a gradient-check test for mini-batch?

**Comment by reunanen** [comment]
@nyanp : Ok, I'll try to add a gradient check for minibatch.

Also, I think the code [here](https://github.com/reunanen/tiny-cnn/blob/80272e443b10c3f8ed53d6d28da054d38aee00b8/tiny_cnn/network.h#L612-L624) should be changed to now run in parallel `n_threads` minibatches – one on each thread. Any comments or objections?

**Comment by nyanp** [comment]
@reunanen
Thanks!

> I think the code here should be changed to now run in parallel n_threads minibatches – one on each thread. Any comments or objections?

I agree. We don't need this parallelism any more. Would you change it?

**Comment by reunanen** [comment]
@nyanp : Sure, I'll have a look at it.

**Comment by reunanen** [comment]
@nyanp
In fact, running the minibatches in parallel would change the learning results. Why? In the current version, the weights are updated after each minibatch, so results from minibatch 1 are already in use when processing minibatch 2. In my opinion, now that I think of it, changing this behavior is a major modification that may well break existing applications. So after all I think it should be introduced in another PR, or perhaps not at all.

**Comment by nyanp** [comment]
@reunanen 
Thank you very much for pointing out the error!

>  In the current version, the weights are updated after each minibatch, so results from minibatch 1 are already in use when processing minibatch 2.

Hmm all weights are updated at [outside of parallel-for loop](https://github.com/nyanp/tiny-cnn/blob/master/tiny_cnn/network.h#L681), so weight should be unchanged within minibatch. Could you tell me where did you find the data-dependency?

Anyway I agree that this issue should be separated from this PR, so I'll merge it sooner.

**Comment by reunanen** [comment]
@nyanp 
Yes, I meant that the current implementation is correct, while running multiple minibatches in parallel (as I [suggested](https://github.com/nyanp/tiny-cnn/pull/181#issuecomment-226988470)) might not be such a great idea after all.

**Comment by nyanp** [comment]
@reunanen
Ah I misunderstood. Ok I'll merge this PR as is :)
